dkg/pedersen/handler.go

```diff
@@ -335,81 +313,86 @@ func (h *Handler) sendDeals(ctx context.Context, out mino.Sender,
 			),
 		)
 
-		dela.Logger.Trace().Msgf("%s sent deal %d", h.me, i)
-		errch := out.Send(dealMsg, participants[i])
+		to := h.startRes.participants[i]
 
-		// this should be further improved by using a worker pool,
-		// as opposed to an unbounded number of goroutines,
-		// but for the time being is okay-ish. -- 2022/08/09
-		go func(errs <-chan error, idx int) {
-			err := <-errs
-			if err != nil {
-				dela.Logger.Warn().Msgf(
-					"got an error while sending deal %v: %v", idx, err)
-				errors <- err
-			} else {
-				done <- idx
-			}
-		}(errch, i)
+		h.log.Info().Str("to", to.String()).Msg("send deal")
+
+		errs := out.Send(dealMsg, to)
+
+		// this can be blocking if the recipient is not receiving message
+		err = <-errs
+		if err != nil {
+			h.log.Err(err).Str("to", to.String()).Msg("failed to send deal")
+		}
 	}
 
-	sent := 0
-	for sent < len(deals) {
-		select {
-		case idx := <-done:
-			dela.Logger.Trace().Msgf("%s sent deal %v", h.me, idx)
-			sent++
+	return nil
+}
 
-		case err := <-errors:
-			dela.Logger.Error().Msgf("%s failed sending a deal: %v", h.me, err)
-			return xerrors.Errorf("failed sending a deal: %v", err)
+func (h *Handler) respond(deals *list.List, out mino.Sender) {
+	numReceivedDeals := 0
 
-		case <-ctx.Done():
-			dela.Logger.Error().Msgf("%s timed out while sending deals", h.me)
-			return xerrors.Errorf("timed out while sending deals")
+	for numReceivedDeals < len(h.startRes.participants)-1 {
+		h.Lock()
+		deal := deals.Front()
+		if deal != nil {
+			deals.Remove(deal)
 		}
-	}
+		h.Unlock()
```

Comment:
I'm still very uncomfortable with this whole setup.
I'd at least consider creating a `syncQueue` structure that wraps around this logic, so that the DKG doesn't have to handle the lock itself, but it gets hidden behind a `.Pop()` call of some kind.